src/actions/CategoriesActions.ts

```diff
@@ -255,3 +257,99 @@ export const defaultCategories = [
   'Transfer:Mycelium',
   'Transfer:Dark Wallet'
 ]
+
+/**
+ * Given an EdgeTxAction, returns the display value for pre-filling the
+ * 'Category' and 'Notes' tiles, if they are not already user-modified.
+ */
+export const getTxActionDisplayInfo = (
+  tx: EdgeTransaction,
+  wallet: EdgeCurrencyWallet,
+  tokenId?: string
+): { splitCategory: SplitCategory; notes?: string } | undefined => {
+  const action = tx.action
+  if (action == null) return
+  const { type } = action
+
+  const getCurrencyCodes = (assets: EdgeAssetAmount[]) => assets.map(asset => getCurrencyCode(wallet, asset.tokenId))
+
+  switch (type) {
+    case 'swap':
+    case 'swapOrderFill': {
+      // Determine if the swap destination was to a different asset or if the
+      // swap source was from a different asset.
+      const txSrcSameAsset = action.sourceAsset.tokenId === tokenId && action.sourceAsset.pluginId === wallet.currencyInfo.pluginId
+      const toFromStr = txSrcSameAsset ? lstrings.transaction_details_swap_to_subcat_1s : lstrings.transaction_details_swap_from_subcat_1s
+      const otherAsset = txSrcSameAsset ? action.destAsset : action.sourceAsset
+
+      return {
+        splitCategory: {
+          category: 'exchange',
+          subcategory: sprintf(toFromStr, getCurrencyCode(wallet, otherAsset.tokenId))
+        }
+      }
+    }
+    case 'swapOrderPost':
+      return {
+        splitCategory: {
+          category: 'expense',
+          subcategory: sprintf(lstrings.transaction_details_swap_order_post)
+        }
+      }
+    case 'swapOrderCancel':
+      return {
+        splitCategory: {
+          category: 'expense',
+          subcategory: sprintf(lstrings.transaction_details_swap_order_cancel)
+        }
+      }
+    case 'stake': {
+      let subcategory
+      if (action.stakeAssets.length === 1) subcategory = sprintf(lstrings.transaction_details_stake_subcat_1s, ...getCurrencyCodes(action.stakeAssets))
+      else if (action.stakeAssets.length === 2) subcategory = sprintf(lstrings.transaction_details_stake_subcat_2s, ...getCurrencyCodes(action.stakeAssets))
+      else {
+        console.error(`Unsupported number of assets for '${type}' EdgeTxActionSwapType`)
```

Comment:
I would use warn instead of error, since as a dev I don't want a big red screen of death if my wallet gets into this state.